soup/man/man1/chksubmit.1
.\" section 1 man page for chksubmit
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2025.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec mis-features and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH chksubmit 1 "28 August 2025" "chksubmit" "IOCCC tools"
.SH NAME
.B chksubmit
\- check an IOCCC submission directory via
.B chkentry \-S
.SH SYNOPSIS
.B chksubmit
.RB [\| \-h \|]
.RB [\| \-v
.IR level \|]
.RB [\| \-V \|]
.RB [\| \-C
.IR chkentry \|]
.RB [\| \-q \|]
.I submission_dir
.SH DESCRIPTION
.BR chksubmit (1)
This tool checks if a IOCCC submission directory will satisfy
most of the important validation checks.
The IOCCC submission directory is specified by
.B submission_dir
argument.
.PP
This tool calls
.B "chkentry \-S"
to perform important validation checks.
.PP
While of this tool does not guarantee that your submission will win the IOCCC,
submitting something that fails to pass the important validation checks
.B will guarantee
that your submission will be
.BR rejected .
.PP
If the
tool is not installed in the default location,
and/or that tool is not found, the path to may be specified
via the
.B \-C\  chkentry
option.
.PP
Unlike
.BR chkentry (1),
this tool is meant to be used by contestants submitting to the IOCCC.
This tool is also is directly invoked by the
.BR mkiocccentry (1)
tool.
.SH OPTIONS
.TP
.B \-h
Print help and exit.
.TP
.BI \-v\  level
Set verbosity level to
.I level
(def: 0).
.TP
.B \-V
Print version and exit.
.TP
.BI \-C\  chkentry
Set the path of the
.BR chkentry (1)
tool to
.BR chkentry .
.TP
.B \-q
Set quiet mode.
Silences msg(), warn(), warnp() if verbosity level is 0.
.SH EXIT STATUS
.TP
0
all OK
.TQ
1
.B chkentry
exited 1 due to some submission directory check failure
.TQ
2
.B \-h
and help string printed or
.B \-V
and version string printed
.TQ
3
command line error
.TQ
>=10
internal error
.SH BUGS
.PP
While the humans maintaining this code use a test suite to validate this code,
and no so-called AI tool contributed to this tool's codebase,
this tool's test suite was written by humans with the best of intentions. :)
.PP
If you have an issue with the tool you can open an issue at:
.sp
.RS 4
.ft B
https://github.com/ioccc\-src/mkiocccentry/issues
.ft R
.RE
.SH EXAMPLES
.PP
Run the tool on an IOCCC submission located in the directory
.IR ./test_ioccc/workdir/test\-1 :
.sp
.RS 4
.ft B
chksubmit ./test_ioccc/workdir/test\-1
.ft R
.RE
.SH SEE ALSO
.PP
.BR mkiocccentry (1),
.BR chkentry (1)


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 2.6.2 2025-09-04

Make adjustments to the new `test_file_util.c` so that it will function in cases
where someone has a file called `foobar` in their directory (this would be me as
I use it as part of testing mkiocccentry).

A man page was started but not added yet because the exit codes in the test code
need to be fixed first and that cannot happen now.

Updated `FILE_UTIL_TEST_VERSION` to `"2.0.1 2025-09-04"`.


## Release 2.6.1 2025-09-02

Moved nearly all of the tree and file utility functions
that were in `jparse/util.c` and `jparse/util.h` over to the
files `soup/file_util.c` and `soup/file_util.h`.
The code that was moved was not needed for the JSON
parser nor the parser's related utilities.
Moving that code will both simplify the JSON
parser code base and make it easier to improve and
enhance tree and file utility functions.
Only a few small utility functions such as `exists()`,
`is_file()`, and `is_read()` remain behind as private
functions in `jparse.l` and `jparse/test_jparse/jnum_gen.c`.

Moved the `jparse/test_jparse/util_test` code over to
as `test_ioccc/test_file_util.c` and ran via the
`test_ioccc/ioccc_test.sh` test script.  That test script
is in its own file and thus no longer has to be copied
from some other source file in order to be compiled.

Restored some of the double blank lines separating functions
in `jparse/util.c`.

The `jparse/test_jparse/run_jparse_tests.sh` no longer runs `util_test`.
Changed the `jparse/test_jparse/jnum_chk` test exit code to `27` to be in sequence.

Made `.gitignore` and `jparse/.gitignore` much more robust, ignoring
intermediate files that may be created while performing `make release`.

Removed `UTIL_TEST` from the `jparse/` tree.

Removed `UTIL_TEST_VERSION` from the `jparse/` tree..  That symbol
has become, in `soup/version.h`, the symbol `FILE_UTIL_TEST_VERSION`
and now has the value "2.0.0 2025-09-01".

Updated, improved and sorted `.gitignore` and `jparse/.gitignore`.

Updated `RUN_JPARSE_TESTS_VERSION` to "2.0.1 2025-09-01".

Updated `JPARSE_UTILS_VERSION` to "2.1.0 2025-09-01".

Updated `JPARSE_REPO_VERSION` to "2.3.0 2025-09-01".

**NO** JSON parser code was changed as a result of
these moves.  Nevertheless the movement of code
serves as a "_checkpoint_" tag and thus warrants
a 2nd-level version number change.


## Release 2.6.0 2025-09-01

Updated `MKIOCCCENTRY_REPO_VERSION` to "2.6.0 2025-09-01"
in anticipation of work that may be needed for **IOCCC29**,
and to tag this version as the starting code base for
**IOCCC29** related development.

Updated `MIN_MKIOCCCENTRY_VERSION` to `MKIOCCCENTRY_VERSION`
which is currently "2.1.2 2025-08-30".

Improved `txzchk_test(8)` with warnings about changing `-d txzchk_tree`.

Fixed echo of `test_ioccc/ioccc_test.sh` command line within
the `test_ioccc/ioccc_test.sh` script.

Synced [jparse repo](https://github.com/xexyl/jparse) and fixed make depend.


## Release 2.5.2 2025-08-30

Fix bugs in txzchk and mkiocccentry.

Although it mostly matters for `islower()` and `isupper()` it is safer to check
`isascii()` before passing it on to one of the `is*()` functions. This is
because of how some implementations do the functions. This probably does not
matter much but it's now safer.

Be more aggressive in errors with txzchk. Although it's still a warning for most
issues with the tarball if a NULL pointer is encountered rather than skipping to
the next line it is an error.

Fix assumption about `find_utils()` in mkiocccentry and txzchk. Before freeing
the `char *` of each tool, also check that it was found. The calls to
`find_utils()` do (for the tools needed) do pass non-NULL and if the tool is
already specified at the command line it still strdup()s it. But the two tools
now check `found_foo` before freeing the strings.

Updated `MKIOCCCENTRY_VERSION` to `"2.1.2 2025-08-30"`.
Updated `TXZCHK_VERSION` to `"2.0.6 2025-08-30"`.

Fixed debug typo in `test_ioccc/chksubmit_test.sh` about a test that should fail.

Changed `test_ioccc/chksubmit_test.sh` IOCCC_TEST_VERSION from "2.1.0 2025-08-28"
to "2.1.1 2025-08-31".

Synced with the [dbg repo](https://github.com/lcn2/dbg), including new git tags.

Synced with the [dyn_array repo](https://github.com/lcn2/dyn_array), including new git tags.

Fixed dependencies in `jparse/test_jparse/Makefile`.

Synced `jparse/version.h` with the [jparse repo](https://github.com/xexyl/jparse) `version.h`.

Synced `jparse/CHANGES.md` with the [jparse repo](https://github.com/xexyl/jparse) `CHANGES.md`.

Updated jparse/CHANGES.md` as needed.


## Release 2.5.1 2025-08-28

Add new tool `chksubmit(1)` as part of issue #1299. This tool runs `chkentry -S`
(new option, see below) and `mkiocccentry(1)` runs it directly instead of
`chkentry`. The `-C` option to `mkiocccentry(1)` now specifies the path to the
new `chksubmit(1)` tool.

Two new options were added to `chkentry(1)` but neither are implemented because
the details are not yet finalised. This is `-S` for `submission mode` and `-s`
for special mode.

The new tool has a man page, `chksubmit.1`.

The template JSON files had major updates due to the update of the `info` and
`auth` structs. Thus the version of these also have been updated.

The test `txzchk` files had to be updated as well due to call to the updated
`find_utils()` function.

Fixed display bug in `mkiocccentry(1)` (inconsistency with a value checked and
what is reported).

Updated the `forbidden_filenames` array. It is not clear if this is complete
yet.

Updated `MKIOCCCENTRY_VERSION` to `"2.1.1 2025-08-28"`.
Updated `INFO_VERSION` to `"1.18 2025-05-28"`.
Updated `AUTH_VERSION` to `"1.24 2025-08-28"`.
Updated `TXZCHK_VERSION` to `"2.0.6 2025-08-28"`.
Updated `CHKENTRY_VERSION` to `"2.1.0 2025-08-28"`.
Updated `gen_test_JSON.sh` version to `"1.0.4 2025-08-28"`.
Set `CHKSUBMIT_VERSION` to `"1.0.0 2025-08-28"`.

The `chksubmit(1)` tool now passes `-S` to `chkentry(1)`.


chksubmit.c
/*
 * chksubmit - check submission directory via chkentry -S
 *
 * "Because grammar and syntax alone do not make a complete language." :-)
 *
 * Copyright (c) 2025 by Cody Boone Ferguson and Landon Curt Noll.
 * All Rights Reserved.
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright, this permission notice and text
 * this comment, and the disclaimer below appear in all of the following:
 *
 *       supporting documentation
 *       source copies
 *       source works derived from this source
 *       binaries derived from this source or from derived source
 *
 * THE AUTHORS DISCLAIM ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
 * ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
 * AUTHORS BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
 * DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * This tool and the JSON parser were co-developed in 2022-2025 by Cody Boone
 * Ferguson and Landon Curt Noll:
 *
 *  @xexyl
 *	https://xexyl.net		Cody Boone Ferguson
 *	https://ioccc.xexyl.net
 * and:
 *	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
 *
 * "Because sometimes even the IOCCC Judges need some help." :-)
 *
 * The concept of the JSON semantics tables was developed by Landon Curt Noll.
 *
 * "Share and Enjoy!"
 *     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
 */


/* special comments for the seqcexit tool */
/* exit code out of numerical order - ignore in sequencing - ooo */
/* exit code change of order - use new value in sequencing - coo */


#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <limits.h>

/*
 * chksubmit - run chkentry -S
 */
#include "chksubmit.h"


#define REQUIRED_ARGS (1)	/* number of required arguments on the command line */

/* number of bytes to hold an int converted to decimal */
#if !defined(CHAR_BIT)
# define CHAR_BIT (8)		/* paranoia - in case limits.h is very old */
#endif
#define INT_DECIMAL_SIZE ((((((size_t)CHAR_BIT) * sizeof(int)) * 302) / 1000) + 1)


/*
 * globals
 */
static bool quiet = false;		    /* true ==> quiet mode */


/*
 * usage message
 *
 * Use the usage() function to print the usage_msg([0-9]?)+ strings.
 */
static const char * const usage_msg =
    "usage: %s [-h] [-v level] [-V] [-C chkentry] [-q] submission_dir\n"
    "\n"
    "\t-h\t\tprint help message and exit\n"
    "\t-v level\tset verbosity level (def level: 0)\n"
    "\t-V\t\tprint version string and exit\n"
    "\t-C chkentry\tpath to chkentry\n"
    "\t-q\t\tquiet mode (def: loud :-) )\n"
    "\t\t\t    NOTE: -q will also silence msg(), warn(), warnp() if -v 0\n"
    "\n"
    "\tsubmission_dir\tsubmission directory to be checked\n"
    "\n"
    "Exit codes:\n"
    "    0\t\tall is OK\n"
    "    1\t\tchkentry exited 1 due to some submission directory check failure\n"
    "    1\t\ta JSON file is not valid JSON, a semantics test failed or some other test failed\n"
    "    2\t\t-h and help string printed or -V and version string printed\n"
    "    3\t\tcommand line error\n"
    "    >=10\tinternal error\n"
    "\n"
    "%s version: %s\n"
    "jparse utils version: %s\n"
    "jparse UTF-8 version: %s\n"
    "jparse library version: %s";


/*
 * functions
 */
static void usage(int exitcode, char const *prog, char const *str) __attribute__((noreturn));
static void chksubmit_sanity_chks(char const *chkentry);
/*
 * usage - print usage to stderr
 *
 * Example:
 *      usage(3, program, "wrong number of arguments");
 *
 * given:
 *	exitcode        value to exit with
 *	str		top level usage message
 *	prog		our program name
 *
 * NOTE: We warn with extra newlines to help internal fault messages stand out.
 *       Normally one should NOT include newlines in warn messages.
 *
 * This function does not return.
 */
static void
usage(int exitcode, char const *prog, char const *str)
{
    /*
     * firewall
     */
    if (prog == NULL) {
	prog = CHKSUBMIT_BASENAME;
	warn(__func__, "\nin usage(): prog was NULL, forcing it to be: %s\n", prog);
    }
    if (str == NULL) {
	str = "((NULL str))";
	warn(__func__, "\nin usage(): str was NULL, forcing it to be: %s\n", str);
    }

    /*
     * print the formatted usage stream
     */
    if (*str != '\0') {
	fprintf_usage(DO_NOT_EXIT, stderr, "%s\n", str);
    }
    fprintf_usage(exitcode, stderr, usage_msg, prog,
	    CHKSUBMIT_BASENAME, CHKSUBMIT_VERSION, JPARSE_UTILS_VERSION, JPARSE_UTF8_VERSION, JPARSE_LIBRARY_VERSION);
    exit(exitcode); /*ooo*/
    not_reached();
}


/*
 * chksubmit_sanity_chks - perform basic sanity checks
 *
 * We perform basic sanity checks on chkentry path.
 *
 * given:
 *
 *      chkentry        - path to chkentry
 *
 * NOTE: this function does not return on error or if things are not sane.
 *
 * NOTE: it is probably wisest to let find_utils() do its thing rather than
 * specify the path with -C.
 */
static void
chksubmit_sanity_chks(char const *chkentry)
{
    /*
     * firewall
     */
    if (chkentry == NULL) {
	err(12, __func__, "called with NULL chkentry");
	not_reached();
    }

    if (!exists(chkentry)) {
        fpara(stderr,
              "",
              "We cannot find chkentry.",
              "",
              "The chkentry program is required to run this tool.",
              "",
              "Perhaps you need to use:",
              "",
              "    chksubmit -C /path/to/chkentry [...]",
              "",
              "or install the mkiocccentry toolkit? You can find the mkiocccentry toolkit at:",
              "",
              "",
              NULL);
        err(13, __func__, "chkentry does not exist: %s", chkentry);
        not_reached();
    }
    if (!is_file(chkentry)) {
        fpara(stderr,
              "",
              "The chkentry path, while it exists, is not a regular file.",
              "",
              "Perhaps you need to use another path:",
              "",
              "    chksubmit -C /path/to/chkentry [...]",
              "",
              "or install the mkiocccentry toolkit?",
              "",
              NULL);
        err(14, __func__, "chkentry is not a regular file: %s", chkentry);
        not_reached();
    }

    if (!is_exec(chkentry)) {
        fpara(stderr,
              "",
              "The chkentry path, while it is a file, is not executable.",
              "",
              "We suggest you check the permissions on the chkentry program, or use another path:",
              "",
              "    chksubmit -C /path/to/chkentry [...]",
              "",
              "or install the mkiocccentry toolkit.",
              "",
              NULL);
        err(15, __func__, "chkentry is not an executable program: %s", chkentry);
        not_reached();
    }

    return;
}


int
main(int argc, char *argv[])
{
    char const *program = NULL;		/* our name */
    char *submission_dir = NULL;        /* directory from which files are to be checked */
    extern char *optarg;		/* option argument */
    extern int optind;			/* argv index of the next arg */
    int exit_code = 0;                  /* return value of chkentry -S */
    bool found_chkentry = false;        /* for find_utils */
    char *chkentry = NULL;              /* for find_utils */
    char v_str[INT_DECIMAL_SIZE+1+1];	/* verbosity level as a string + NUL + 1 for paranoia */
    int i = 0;

    /* IOCCC requires use of C locale */
    set_ioccc_locale();

    /*
     * parse args
     */
    program = argv[0];
    while ((i = getopt(argc, argv, ":hv:VqC:")) != -1) {
	switch (i) {
	case 'h':		/* -h - print help to stderr and exit 0 */
	    usage(2, program, "");	/*ooo*/
	    not_reached();
	    break;
	case 'v':		/* -v verbosity */
	    /*
	     * parse verbosity
	     */
	    verbosity_level = parse_verbosity(optarg);
	    if (verbosity_level < 0) {
		usage(3, program, "invalid -v verbosity"); /*ooo*/
		not_reached();
	    }
	    break;
	case 'V':		/* -V - print version and exit */
	    print("%s version: %s\n", CHKSUBMIT_BASENAME, CHKSUBMIT_VERSION);
	    print("jparse utils version: %s\n", JPARSE_UTILS_VERSION);
	    print("jparse UTF-8 version: %s\n", JPARSE_UTF8_VERSION);
	    print("jparse library version: %s\n", JPARSE_LIBRARY_VERSION);
	    exit(2);		/*ooo*/
	    not_reached();
	    break;
        case 'C': /* -C chkentry */
            chkentry = optarg;
            break;
	case 'q':
	    quiet = true;
	    msg_warn_silent = true;
	    break;
	case ':':   /* option requires an argument */
	case '?':   /* illegal option */
	default:    /* anything else but should not actually happen */
	    check_invalid_option(program, i, optopt);
	    usage(3, program, ""); /*ooo*/
	    not_reached();
	    break;
	}
    }

    /*
     * we need chkentry if the user did not specify the path
     */
    if (chkentry == NULL) {
        find_utils(NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &found_chkentry, &chkentry,
                NULL, NULL, NULL, NULL, NULL, NULL);
    }

    /*
     * verify chkentry is an executable program
     */
    chksubmit_sanity_chks(chkentry);

    argc -= optind;
    argv += optind;
    switch (argc) {
    case 1:
        submission_dir = argv[0];
	break;
    case 2:
        if (!strcmp(argv[0], ".") && !strcmp(argv[1], ".")) {
            vrergfB(-1, -1); /* Easter egg */
            not_reached();
        }
        /*fallthrough*/
    default:
	usage(3, program, "wrong number of arguments");	/*ooo*/
	not_reached();
	break;
    }

    /*
     * debugging
     */
    if (dbg_allowed(DBG_MED)) {
        if (quiet) {
            dbg(DBG_MED, "about to execute: %s -S -v %d -q -- %s", chkentry, verbosity_level, submission_dir);
        } else {
            dbg(DBG_MED, "about to execute: %s -S -v %d -- %s", chkentry, verbosity_level, submission_dir);
        }
    }

    /*
     * execute chkentry -S [-v level] [-q]
     */
    if (snprintf(v_str, sizeof(v_str)-2, "%d", verbosity_level) <= 0) {
	err(16, program, "failed to convert int: %d into decimal", i);
	not_reached();
    }
    v_str[sizeof(v_str)-1] = '\0';	/* paranoia */
    if (quiet) {
	if (is_dbg_enabled) {
	    exit_code = shell_cmd(__func__, false, true, "% -S -v % -q -- %", chkentry, v_str, submission_dir);
	} else {
	    exit_code = shell_cmd(__func__, false, true, "% -S -q -- %", chkentry, submission_dir);
	}
    } else {
	if (is_dbg_enabled) {
	    exit_code = shell_cmd(__func__, false, true, "% -S -v % -- %", chkentry, v_str, submission_dir);
	} else {
	    exit_code = shell_cmd(__func__, false, true, "% -S -- %", chkentry, submission_dir);
	}
    }

    /*
     * free storage
     */
    if (chkentry != NULL && found_chkentry) {
        free(chkentry);
        chkentry = NULL;
    }

    /*
     * All Done!!! All Done!!! -- Jessica Noll, Age 2
     *
     */
    if (exit_code != 0) {
        exit(1); /*ooo*/
        not_reached();
    }
    exit(0); /*ooo*/
}
